src/Workspaces/Core/Portable/Workspace/Solution/DocumentInfo.cs

```diff
@@ -181,50 +193,60 @@ internal class DocumentAttributes : IChecksummedObject, IObjectWritable
                 IEnumerable<string> folders = null,
                 Optional<SourceCodeKind> sourceCodeKind = default(Optional<SourceCodeKind>),
                 Optional<string> filePath = default(Optional<string>),
+                Optional<string> projectFilePath = default(Optional<string>),
                 Optional<bool> isGenerated = default(Optional<bool>))
             {
                 var newId = id ?? Id;
                 var newName = name ?? Name;
                 var newFolders = folders ?? Folders;
                 var newSourceCodeKind = sourceCodeKind.HasValue ? sourceCodeKind.Value : SourceCodeKind;
                 var newFilePath = filePath.HasValue ? filePath.Value : FilePath;
+                var newProjectFilePath = projectFilePath.HasValue ? projectFilePath.Value : ProjectFilePath;
                 var newIsGenerated = isGenerated.HasValue ? isGenerated.Value : IsGenerated;
 
                 if (newId == Id &&
                     newName == Name &&
                     newFolders == Folders &&
                     newSourceCodeKind == SourceCodeKind &&
                     newFilePath == FilePath &&
+                    newProjectFilePath == ProjectFilePath &&
                     newIsGenerated == IsGenerated)
                 {
                     return this;
                 }
 
-                return new DocumentAttributes(newId, newName, newFolders, newSourceCodeKind, newFilePath, newIsGenerated);
+                return new DocumentAttributes(newId, newName, newFolders, newSourceCodeKind, newFilePath, newProjectFilePath, newIsGenerated);
             }
 
             public void WriteTo(ObjectWriter writer)
             {
+                // these information is volatile. it can be different
+                // per session or not content based value. basically not
+                // persistable. these information will not be included in checksum
                 Id.WriteTo(writer);
 
+                writer.WriteString(FilePath);
```

Comment:
> so your concern is someone might save data which is not actually persistable but doing it without knowing it?

Yes.  Of course :)   That's how all the new IDE features are working.  So if their checksums are nto actually persistable that's a problem :)

> that is already happening in current system. there is nothing that prevent people saving data which is not persistable.

That's a problem :)

> ut, the one who creates the checksum automatically favoring persistable checksum seems better system.

Agreed.  But right now that's not what this PR is doing.  it seems to be silently failing over to non-persistable checksums, and the caller has no idea that that's the case.